### ISSUE TITLE
[Feature] remove the GaleraAgent sidecar from the MariaDB pods

### ISSUE THREAD

**Issue by businessbean** [mariadb-operator/mariadb-operator#785]
## Is your feature request related to a problem? Please describe

Currently, all MariaDB pods will be restarted after the operator has been updated. This is caused by the version change for `GaleraAgent` sidecar, because it uses the operator binary for the agent functions. In a Kubernetes cluster with multiple MariaDB Galera cluster instances requested for different services, all of them will be impacted shortly after the operator update. At least it will cause additional Galera IST I/Os and depending on the behavior of the service also short outages for the users.

## Describe the solution you'd like

Per requested `mariadbs.k8s.mariadb.com` object one dedicated `GaleraAgent` pod should be deployed. Optionally for every `mariadbs` pod a `GaleraAgent` pod can be deployed (to provide HA). I would avoid moving the agent functionality into the operator pod itself, because with a lot of `mariadbs.k8s.mariadb.com` instances it can become a SPOF or performance bottleneck.

The drawback of the dedicated `GaleraAgent` pods will be that socket connections to the database are not possible anymore, but i think this is a minor issue compared to the database restarts.

## Describe alternatives you've considered

A dedicated GaleraAgent binary will maybe help, but if code is changed in that area you need again to restart the DB pods.

**Environment details**:
- Kubernetes version: `1.30.3`
- Kubernetes distribution: `Vanilla`
- mariadb-operator version: `0.0.29`
- Install method: `n/a`
- Install flavor: `n/a`

**Comment by mmontes11** [comment]
Hey there @businessbean ! Thanks for bringing this up.

I think there is a misunderstanding here on how the operator works.

The operator has some default images, including the agent and init ones, which are bumped on every operator release and are used for defaulting. When updating the operator, if you have a `MariaDB` already running, the operator will preserve the image versions used for the init and agent containers. For this reason, whenever we release a feature in the operator that depends on new changes on the init or agent containers, we always encourage users to update, as we did in [`v0.0.30`](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPGRADE_v0.0.30.md):
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPGRADE_v0.0.30.md 

The user is in control of the mariadb, agent and init container versions, the operator does not update image version unless the user tolds it to do so!

That said, sometimes changes in the `Pod` template are introduced, leading to `MariaDB` rolling updates when the operator gets updated. We are aware of this, and to mitigate this, we have introduced the ability to `suspend` upgrades on `MariaDBs`. This way, you are able to selectively roll out  all instances when updating the operator:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/SUSPEND.md

I hope it is more clear now, but let me know if you have further questions!

**Comment by mmontes11** [comment]
Hey there @businessbean !

Thank you very much for your input. Here my answer for each of the points you made:

Regarding 1,  that's correct, `suspend` would pause all reconciliation, not only upgrades but also cluster recovery for instance. As an alternative, you can use the `OnDelete` `updateStrategy` documented [here](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPDATES.md#ondelete), which will only perform upgrades when the `Pods` get deleted. If the `Pods` get restarted, they will also get updated, so we could implement an `updateStrategy.type = Never`, which would make the operator to `Never` update the `Pod` template and still perform other operations as usual, like the Galera cluster recovery. Let me know if you are interested in this feature and we can make some room for it in the roadmap.

Regarding 2, Our current Galera implementation relies on both an `init` and a `galera` containers which run with the operator image. Whilst the `galera` container could run as a separate `Deployment` , the `init` container still needs to run before the `mariadb` container comes up. Since the `init` container is also part of the `MariaDB` `Pod` template, there is nothing we can do to decouple the operator containers from the DB ones. For making this happen we would need to do a complete re-architecture of the Galera setup, ditching the `StatefulSet` and managing the `Pods` from the operator. This would take a massive amount of time, it is less standard and therefore also harder to maintain.

That said, as you could see we are ironing out our Galera support to be as production ready as possible, that's the reason why we are introducing changes in the `init` and `galera` containers in the last releases. Once we've reached a stable stage, I'm not expeting major changes in these containers unless there is a bug or a feature that requires operations on the `MariaDB` `Pod` side, something we try to avoid, we opt for doing operations on the operator side whenever possible. 

In conclusion, I think running the `mariadb-operator` containers alongside the `MariaDB`  ones is an affordable trade-off we are making, having a complete custom architecture would provide us with more flexibility but the development and maintenance cost of that would be huge compared to having to infrequently update `mariadb-operator` containers on the `MariaDB` `Pod`. This isn't something new, it is a pattern followed by other database operators and service mesh projects. They call these additional containers the data-plane, and we could actually explore how to make data-plane upgrades better!

Let me know what you think!
